README.md
# ddp-proxy

Manage multiple Meteor server to server DDP connections and provides a few
features on top of the standard `DDP.connect()`.

## Features

- Expire inactive connections
- Built-in login mechanism
- Resume connections based on session key and login credentials
- Automatic creation of required Mongo collections when subscribing

## Install

To create DDP proxy connections,

```
meteor add xamfoo:ddp-proxy
```

To accept DDP proxy connections,

```
meteor add xamfoo:ddp-proxy-endpoint
```

`ddp-proxy-endpoint` package prevents auto-published data from sending every
time a new DDP proxy connection is made. This reduces the network overhead if
the auto-published information is not always required.

### Dependencies
- `underscore`
- `ejson`
- `meteor`
- `mongo`
- `ddp`

## Getting Started

```javascript
var ddpProxy = new DDPProxy({url: 'http://localhost:3000'});
var connection = ddpProxy.connect();
connection.call('addFruit', function (err, res) { // Call a method
  console.log('method called', err, res);
});
connection.subscribe('fruits', function () { // Subscribe
  // Results are returned in connection.collections
  console.log(connection.collections.fruits.find().fetch());
});
```

## Usage

### `new DDPProxy([options])`

DDPProxy class

Parameters:

- {object} options - Configuration object. Refer to configure()
- {Mongo.Collection} options.collection[=localCollection] -
  Mongo collection used to store connection data. If not specified, connection
  info will be stored in memory.

### `ddpProxy.configure([options])`

Configure options

Parameters:

- {object} options
- {string} options.url - Default url for connection if not specified.
  Set to Meteor.absoluteUrl() if not specified.
- {number} options.connectionExpire[=900] - Time(seconds) before a
  non-session connection expires. Default is 15 minutes.
- {number} options.connectionTimeout[=10] - Time(seconds) before a
  connection attempt timeouts. Default is 10 seconds.
- {number} options.sessionExpire[=900] - Time(seconds) before a
  connection with a session id expires. Default is 15 minutes.
- {number} options.sessionExpireOnResume[=900] - The new expiry
  time(seconds) to set on the connection when a session is resumed. Default
  is 15 minutes.
- {number} options.expireInterval[=120] - Time interval to expire
  connections.
- {function} options.isValidSession - A function for checking
  if a DDP proxy session id is valid. By default it is a function that
  always return true.

### `ddpProxy.connect([options])`

Creates a DDP `proxyConnection` to the given server url and login options

Parameters:

- {object} options
- {string} [options.url] - The connection url
- {\*} [options.sessionId] - Session id which is used to resume this
  connection. A session id must be a string or a serializable object.
- {object} [options.login] - Login options for Meteor
- {object} [options.login.resume] - Resume token for Meteor login
- {boolean} options.autoPublish[=false] - If autoPublish is set to true, the
  connection will subscribe to autopublished data.

Example login with password:

```javascript
var ddpProxy = new DDPProxy;
var connection = ddpProxy.connect({
  login: {
    user: {email: 'test@test.com'},
    password: {
      algorithm: 'sha-256',
      digest: SHA256('test') // SHA256 from Meteor `sha` package
    }
  }
});

// Login with the obtained resume token
var connection2 = ddpProxy.connect({
  login: {
    resume: connection.loginStatus.result.token
  }
});
```

Example of resuming a connection:

```
var ddpProxy = new DDPProxy;
var connection = ddpProxy.connect({sessionId: 'mysession'});

var resumeConnection = ddpProxy.connect({sessionId: 'mysession'});
resumeConnection === connection // true
```

### `ddpProxy.startExpire()`

Start monitoring connections and remove expired connections.

### `ddpProxy.stopExpire()`

Stop monitoring connections expiry

### `ddpProxy.stop()`

Close all connections and stop monitoring

### `proxyConnection.collections`

An object which contains (collection name, Mongo.Collection) key value pairs
which are created when subscriptions requires them.

Example:

```javascript
// On the target server
Fruits = new Mongo.Collection('fruits');

Meteor.publish('fruits', function () {
  return Fruits.find();
});
```

```javascript
// On the proxy server
var ddpProxy = new DDPProxy;
var connection = ddpProxy.connect();

connection.subscribe('fruits', function () { // Subscribe
  // Results are returned in connection.collections
  console.log(connection.collections.fruits.find().fetch());
});
```

### `proxyConnection.loginStatus`

Example loginStatus object:
```javascript
{
  "result": {
    // User id
    "id": "GiWKSpcXbamqbukZM",
    // User resume token
    "token": "uenZILN21QooIS9rsIXOcJwcl-iKG0puyzLCRR_0Tl4",
    "tokenExpires": {
      "$date": 1435649944694
    }
  }
}
```

Example loginStatus object(error):
```javascript
{
  "error": [errorObject]
}
```


CHANGELOG.md
## 0.2.0

- Support Meteor 1.2.0.1 core packages
- Add autoPublish option to DDPProxy.connect()

## 0.1.0

- First release


ddp-proxy-endpoint/tests.js
Tinytest.add('Stop autopublish for DDP proxy connections', function (test) {
  var proxy = new DDPProxy;
  var connection = proxy.connect();

  Meteor._sleepForMs(1000);
  test.equal(Object.keys(connection.collections).length === 0, true);

  proxy.stop();
});

Tinytest.add('Does not affect normal DDP connections', function (test) {
  var connection = DDP.connect(Meteor.absoluteUrl());
  var Fruits = new Mongo.Collection('fruits', {connection: connection});
  Meteor._sleepForMs(1000);
  test.equal(Fruits.find().count() > 0, true);
  connection.close();
});


ddp-proxy/ddp-proxy-tests.js
var Future = Npm.require('fibers/future');

function subscribeSync (connection /*, arguments*/) {
  var args = Array.prototype.slice.call(arguments, 1);
  var future = new Future;
  connection.subscribe.apply(connection, args.concat({
    onReady: function () { future['return'](); },
    onError: function (e) { future['return'](e); }
  }));
  return future.wait();
}

function methodCallSync (connection /*, arguments*/) {
  var args = Array.prototype.slice.call(arguments, 1);
  var call = Meteor.wrapAsync(connection.call, connection);
  return call.apply(connection, args);
};

var loginOptions = {
  user: {email: 'test@test.com'},
  password: {
    algorithm: 'sha-256',
    digest: SHA256('test')
  }
};

// Set up DDP server listener
var DDPServer = (function () {
  function DDPServer () {
    var self = this;
    self._listeners = [];

    Meteor.server.stream_server.register(function (socket) {
      socket.on('data', function () {
        var socketSelf = this;
        var args = arguments;

        self._listeners.forEach(function (listener) {
          listener.apply(socketSelf, args);
        });
      });
    });
  }
  DDPServer.prototype.addListener = function (fn) {
    return this._listeners.push(fn);
  };
  DDPServer.prototype.removeListener = function (fn) {
    var self = this;
    var index = self._listeners.indexOf(fn);
    if (index >= 0) self._listeners.splice(index, 1);
  };
  return new DDPServer;
}());

Tinytest.add(
  'Indicate ddpproxy support in connection request', function (test) {
    var proxy = new DDPProxy;
    var testCount = 0;
    var listener = Meteor.bindEnvironment(function (raw_msg) {
      try {
        raw_msg = JSON.parse(raw_msg);
        if (raw_msg && raw_msg.msg === 'connect') {
          test.equal(
            raw_msg.support && raw_msg.support.indexOf('ddpproxy') >= 0,
            true
          );
          testCount += 1;
        }
      }
      catch (e) {}
    });
    DDPServer.addListener(listener);
    var connection = proxy.connect();
    test.equal(testCount > 0, true);
    DDPServer.removeListener(listener);

    proxy.stop();
  }
);

Tinytest.add(
  'No ddpproxy support in connection request when autopublish option is true',
  function (test) {
    var proxy = new DDPProxy;
    var testCount = 0;
    var listener = Meteor.bindEnvironment(function (raw_msg) {
      try {
        raw_msg = JSON.parse(raw_msg);
        if (raw_msg && raw_msg.msg === 'connect') {
          test.equal(
            raw_msg.support && !(raw_msg.support.indexOf('ddpproxy') >= 0),
            true
          );
          testCount += 1;
        }
      }
      catch (e) {}
    });
    DDPServer.addListener(listener);
    var connection = proxy.connect({autoPublish: true});
    test.equal(testCount > 0, true);
    DDPServer.removeListener(listener);

    proxy.stop();
  }
);

Tinytest.add('New connection', function (test) {
  var proxy = new DDPProxy();
  var connection = proxy.connect();
  test.equal(connection.status().connected, true);

  proxy.stop();
});

Tinytest.add('Subscribe on connection', function (test) {
  var proxy = new DDPProxy();
  var connection = proxy.connect();
  subscribeSync(connection, 'fruits');
  test.equal(connection.collections.fruits.find().count(), 1);
  test.equal(
    connection.collections.fruits.findOne().name, 'apple'
  );

  proxy.stop();
});

Tinytest.add('Call method on connection', function (test) {
  var proxy = new DDPProxy();
  var connection = proxy.connect();
  subscribeSync(connection, 'fruits');
  methodCallSync(connection, 'removeFruits');
  test.equal(connection.collections.fruits.find().count(), 0);
  methodCallSync(connection, 'addFruit', 'apple');
  test.equal(connection.collections.fruits.find().count(), 1);

  proxy.stop();
});

Tinytest.add('Resume session id', function (test) {
  var proxy = new DDPProxy();
  var connection;

  connection = proxy.connect({sessionId: 'apple'});
  subscribeSync(connection, 'fruits');
  methodCallSync(connection, 'removeFruits');
  methodCallSync(connection, 'addFruit', 'apple');
  test.equal(!!connection.collections.fruits, true);

  connection = proxy.connect();
  test.equal(!!connection.collections.fruits, false);

  connection = proxy.connect({sessionId: 'banana'});
  test.equal(!!connection.collections.fruits, false);

  connection = proxy.connect({sessionId: 'apple'});
  test.equal(!!connection.collections.fruits, true);
  test.equal(connection.collections.fruits.find().count(), 1);

  proxy.stop();
});

Tinytest.add('Resume object session id', function (test) {
  var proxy = new DDPProxy();
  var connection;

  connection = proxy.connect({sessionId: {apple: 1}});
  subscribeSync(connection, 'fruits');
  methodCallSync(connection, 'removeFruits');
  methodCallSync(connection, 'addFruit', 'apple');
  test.equal(!!connection.collections.fruits, true);

  connection = proxy.connect();
  test.equal(!!connection.collections.fruits, false);

  connection = proxy.connect({sessionId: {banana: 1}});
  test.equal(!!connection.collections.fruits, false);

  connection = proxy.connect({sessionId: {apple: 1}});
  test.equal(!!connection.collections.fruits, true);
  test.equal(connection.collections.fruits.find().count(), 1);

  proxy.stop();
});

Tinytest.add('Login with password', function (test) {
  var proxy = new DDPProxy;
  var connection = proxy.connect({login: loginOptions});
  test.equal(
    !!connection.loginStatus && !!connection.loginStatus.result &&
      !!connection.loginStatus.result.token,
    true
  );

  proxy.stop();
});

Tinytest.add('Login with invalid credentials', function (test) {
  var proxy = new DDPProxy;
  var connection = proxy.connect({login: {user: 'invalid@test.com'}});
  test.equal(
    !!connection.loginStatus && !!connection.loginStatus.error,
    true
  );

  proxy.stop();
});

Tinytest.add('Resume login with password', function (test) {
  var proxy = new DDPProxy;
  var connection = proxy.connect({login: loginOptions});
  var token = connection.loginStatus.result.token;
  var resume = proxy.connect({login: {resume: token}});
  test.equal(
    !!connection.loginStatus && !!connection.loginStatus.result &&
      !!connection.loginStatus.result.token,
    true
  );
});

Tinytest.add('Configure options', function (test) {
  var proxy = new DDPProxy;
  var oldConfig = proxy._config;
  var newConfig = {
    ddpConnection: {_dontPrintErrors: true},
    isValidSession: function () { return false; }
  }
  proxy.configure(newConfig);
  Object.keys(proxy._config).forEach(function (k) {
    if (k in newConfig) test.equal(proxy._config[k], newConfig[k]);
    else test.equal(proxy._config[k], oldConfig[k]);
  });

  proxy.stop();
});

Tinytest.add('Invalid url and connectionTimeout option', function (test) {
  var proxy = new DDPProxy(
    {connectionTimeout: 1, ddpConnection: {_dontPrintErrors: true}}
  );
  test.throws(function () {
    proxy.connect({url: 'http://1'});
  }, 'Connection timeout');
  proxy.stop();
});

Tinytest.add('Remove connection on close()', function (test) {
  var proxy = new DDPProxy;
  test.equal(Object.keys(proxy._connections).length, 0);
  test.equal(proxy._cln.find().count(), 0);

  var connection = proxy.connect();
  test.equal(Object.keys(proxy._connections).length, 1);
  test.equal(proxy._cln.find().count(), 1);

  connection.close();
  test.equal(Object.keys(proxy._connections).length, 0);
  test.equal(proxy._cln.find().count(), 0);

  proxy.stop();
});

Tinytest.add('connectionExpire option', function (test) {
  var proxy = new DDPProxy({
    connectionExpire: 0,
  });
  var connection = proxy.connect();
  var data = proxy._cln.findOne();
  test.equal(data.expire <= new Date, true);

  proxy.stop();
});

Tinytest.add('sessionExpire option', function (test) {
  var proxy = new DDPProxy({sessionExpire: 0});
  var connection = proxy.connect({sessionId: 'apple'});
  var data = proxy._cln.findOne({sessionId: 'apple'});
  test.equal(data.expire <= new Date, true);

  proxy.stop();
});

Tinytest.add('sessionExpireOnResume option', function (test) {
  var proxy = new DDPProxy({sessionExpireOnResume: 1234});
  var connection = proxy.connect({sessionId: 'apple'});

  connection = proxy.connect({sessionId: 'apple'});
  var data = proxy._cln.findOne({sessionId: 'apple'});
  test.equal(Math.abs(data.expire - (new Date) - 1234 * 1000) < 100, true);

  proxy.stop();
});

Tinytest.add('isValidSession option', function (test) {
  var proxy = new DDPProxy({isValidSession: function (sessionId) {
    if (sessionId && sessionId.name === 'apple') return true;
    return false;
  }});
  test.equal(proxy.connect({sessionId: '1234'}), undefined);
  test.equal(
    proxy.connect({sessionId: {name: 'apple'}}),
    proxy.connect({sessionId: {name: 'apple'}})
  );

  proxy.stop();
});

Tinytest.add(
  'Expiring connections and expireInterval option', function (test) {
    var proxy = new DDPProxy({expireInterval: 1});
    var connection = proxy.connect();
    proxy._cln.find().forEach(function (doc) {
      proxy._cln.update({_id: doc._id}, {$set: {expire: new Date}});
    });
    test.equal(proxy._cln.find().count(), 1);
    Meteor._sleepForMs(1100);
    test.equal(proxy._cln.find().count(), 0);

    proxy.stop();
  }
);

Tinytest.add('stopExpire()', function (test) {
  var proxy = new DDPProxy({expireInterval: 1});
  proxy.stopExpire();
  var connection = proxy.connect();
  proxy._cln.find().forEach(function (doc) {
    proxy._cln.update({_id: doc._id}, {$set: {expire: new Date(1)}});
  });
  test.equal(proxy._cln.find().count(), 1);
  Meteor._sleepForMs(1100);
  test.equal(proxy._cln.find().count(), 1);

  proxy.stop();
});

Tinytest.add('stop()', function (test) {
  var proxy = new DDPProxy({expireInterval: 1});
  var connection = proxy.connect();
  proxy._cln.find().forEach(function (doc) {
    proxy._cln.update({_id: doc._id}, {$set: {expire: new Date(1)}});
  });
  test.equal(proxy._cln.find().count(), 1);

  proxy.stop();
  test.equal(proxy._cln.find().count(), 0);
});


ddp-proxy/package.js
Package.describe({
  name: 'xamfoo:ddp-proxy',
  version: '0.2.0',
  // Brief, one-line summary of the package.
  summary: 'Connection manager for a DDP proxy server',
  // URL to the Git repository containing the source code for this package.
  git: 'https://github.com/xamfoo/ddp-proxy',
  // By default, Meteor will default to using README.md for documentation.
  // To avoid submitting documentation, set this field to null.
  documentation: 'README.md'
});

Package.onUse(function(api) {
  api.versionsFrom('1.2.0.1');
  api.use([
    'meteor',
    'ejson',
    'ddp',
    'ddp-common',
    'underscore',
    'mongo',
    'random'
  ], 'server');
  api.addFiles(['ddp-proxy.js'], 'server');
  api.export('DDPProxy', 'server');
});

Package.onTest(function(api) {
  api.versionsFrom('1.2.0.1');
  api.use([
    'tinytest',
    'mongo',
    'accounts-password',
    'sha',
  ], 'server');
  api.use('xamfoo:ddp-proxy');
  api.addFiles(['test-fixtures.js', 'ddp-proxy-tests.js'], 'server');
});


ddp-proxy-endpoint/ddp-proxy-endpoint.js
// We trigger the creation of the session object by temporarily creating a
// DDP connection to the server
var tmpConnection = DDP.connect(Meteor.absoluteUrl());

// Not a very clean way, but we poll until we get a reference to a server
// session object
var interval = Meteor.setInterval(function () {
  var SessionKeys = Object.keys(Meteor.server.sessions);
  if (!SessionKeys.length) return;
  var Session = Meteor.server.sessions[SessionKeys[0]].constructor;

  // Wrap startUniversalSubs to check for Session.noAutoPublish
  Session.prototype.startUniversalSubs = (function (start) {
    return function () {
      var self = this;
      Meteor.setTimeout(function () {
        if (self.noAutoPublish) return;
        start.apply(self, arguments);
      }, 0);
    };
  })(Session.prototype.startUniversalSubs);

  // Wrap _handleConnect with another handle
  Meteor.server.constructor.prototype._handleConnect = (function (handle) {
    return function (socket, msg) {
      var self = this;
      handle.apply(self, arguments);

      if (msg && msg.support && msg.support.indexOf('ddpproxy') > -1) {
        socket._meteorSession.noAutoPublish = true;
      }
    };
  })(Meteor.server.constructor.prototype._handleConnect);

  // Make sure to clean up the connection
  tmpConnection.disconnect();
  // Stop polling
  Meteor.clearInterval(interval);
}, 500);


ddp-proxy/ddp-proxy.js
var Future = Npm.require('fibers/future');

// Get Connection constructor and supported DDP versions
var Connection = (function () {
  var connection = DDP.connect(Meteor.absoluteUrl());
  connection.close();
  return connection.constructor;
}());

var parseDDP = DDPCommon.parseDDP;

// Callback for automatic creation of new mongo collections when data is
// received
var onData = function (conn, raw_msg) {
  var msg;

  try {
    msg = DDPCommon.parseDDP(raw_msg);
  } catch (e) {
    return;
  }
  if (!msg || !msg.msg || !msg.collection) return;

  conn.collections[msg.collection] =
    conn.collections[msg.collection] ||
    new Mongo.Collection(msg.collection, {connection: conn});
};

/**
 * Extends the DDP connection class.
 */
var ProxyConnection = (function (_Connection) {
  function ProxyConnection () {
    var self = this;
    _Connection.apply(self, arguments);

    self.collections = self.collections || {};
    self._onClose = [];
    self._stream.on(
      'message',
      Meteor.bindEnvironment(_.partial(onData, self), Meteor._debug)
    );
  }
  ProxyConnection.prototype = Object.create(_Connection.prototype, {
    constructor: {
      value: ProxyConnection,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (Object.setPrototypeOf) Object.setPrototype(ProxyConnection, _Connection);
  else ProxyConnection.__proto__ = _Connection;

  ProxyConnection.prototype.close = function close () {
    var self = this;
    _Connection.prototype.close.apply(self, arguments);
    while (self._onClose.length) { self._onClose.pop()(); }
  };

  ProxyConnection.prototype.onClose = function onClose (func) {
    if (typeof func !== 'function')
      throw new Error('Argument must be a function');
    this._onClose.push(func);
  };

  return ProxyConnection;
}(Connection));

/**
 * DDPProxy class
 *
 * @param {object} options - Configuration object. Refer to configure()
 * @param {Mongo.Collection} options.collection[=localCollection] -
 *   Mongo collection used to store connection data. If not specified,
 *   connection info will be stored in memory.
 */
DDPProxy = function DDPProxy (opt) {
  var self = this;
  opt = opt || {};

  // DDP connections;
  self._connections = {};
  // Mongo collection to store connection state
  self._cln = opt.collection || new Mongo.Collection(null);

  // Id of interval to expire connections
  self._expireIntervalId = null;

  // Configuration
  self._config = {};
  self.configure(opt);
};

_.extend(DDPProxy.prototype, {
  _addConnection: function (opt) {
    var self = this;
    opt = opt || {};
    var url = opt.url;
    var sessionId = opt.sessionId;
    var loginOptions = opt.login;
    var connectFuture = new Future();
    var loginFuture = new Future();
    var connectionTimeout;
    var onConnected = _.partial(function (future) {
      Meteor.clearTimeout(connectionTimeout);
      !future.isResolved() && future['return']();
    }, connectFuture);
    var ddpOption = _.extend({}, self._config.ddpConnection, {
      onConnected: onConnected,
      supportedDDPVersions: !opt.autoPublish ?
        DDPCommon.SUPPORTED_DDP_VERSIONS.concat('ddpproxy') :
        DDPCommon.SUPPORTED_DDP_VERSIONS
    });
    if (self._config.ddpConnection.onConnected) {
      ddpOption.onConnected = _.partial(function (fn) {
        onConnected.apply(this, arguments);
        fn.apply(this, arguments);
      }, self._config.ddpConnection.onConnected);
    }
    var insertId;

    var connection = new ProxyConnection(url, ddpOption);
    connection.onClose(function () {
      delete self._connections[insertId];
      self._cln.remove({_id: insertId});
    });

    connectionTimeout = Meteor.setTimeout(function () {
      connection.close();
      !connectFuture.isResolved() && connectFuture['throw'](
        new Error('Connection timeout')
      );
    }, self._config.connectionTimeout * 1000);

    var doc = {
      url: url,
      sessionId: sessionId,
      resume: null,
      expire: sessionId ?
        new Date((new Date()).getTime() + self._config.sessionExpire * 1000) :
        new Date((new Date()).getTime() + self._config.connectionExpire * 1000)
    };
    var loginCallback = _.partial(
      function (controller, connection, future, err, res) {
        if (err) {
          connection.loginStatus = {error: err};
          controller._removeConnection(insertId);
        }
        else if (!res || !res.token) {
          connection.loginStatus = {error: new Error('No token was generated')};
          controller._removeConnection(insertId);
        }
        else {
          connection.loginStatus = {result: res};
          controller._cln.update({_id: insertId}, {$set: {resume: res.token}});
        }

        !future.isResolved() && future.return();
      }, self, connection, loginFuture
    );

    if (_.size(loginOptions) === 0) {
      insertId = self._cln.insert(doc);
      self._connections[insertId] = connection;
      loginFuture.return();
    }
    else {
      insertId = self._cln.insert(
        _.extend(doc, {resume: Random.id(42)})
      );
      self._connections[insertId] = connection;

      connection.onReconnect = function () {
        connection.call('login', loginOptions, loginCallback);
      };
      connection.reconnect();
    }

    connectFuture.wait();
    loginFuture.wait();

    return connection;
  },

  _removeConnection: function (connectionId) {
    var self = this;
    var connection = self._connections[connectionId];
    if (connection) connection.close();
  },

  _removeExpiredConnections: function () {
    var self = this;
    var now = new Date();
    self._cln.find({expire: {$lt: now}}).forEach(function (doc) {
      self._removeConnection(doc._id);
    });
    return self;
  },

  _resumeConnection: function (url, sessionId, resume) {
    var self = this;
    // resume = typeof resume === 'string' ? Accounts._hashLoginToken()

    var connectionData = self._cln.findOne({
      url: url, resume: resume, sessionId: sessionId,
      expire: {$gte: new Date}
    });
    if (!connectionData || !connectionData._id) return undefined;

    // Update expiry
    self._cln.update(
      {_id: connectionData._id},
      {
        $set: {
          expire: new Date(
            (new Date).getTime() + self._config.sessionExpireOnResume * 1000
          )
        }
      }
    );
    var connection = self._connections[connectionData._id];
    if (!connection || !connection.status) return undefined;

    var status = connection.status().status;
    if (status === 'failed' || status === 'offline') {
      self._removeConnection(connectionData._id);
      return undefined;
    }

    return connection;
  },

  /**
   * Configure options
   *
   * @param {object} options
   * @param {string} options.url -
   *   Default url for connection if not specified.
   *   Set to Meteor.absoluteUrl() if not specified.
   * @param {number} options.connectionExpire[=900] -
   *   Time(seconds) before a non-session connection expires. Default is 15
   *   minutes.
   * @param {number} options.connectionTimeout[=10] -
   *   Time(seconds) before a connection attempt timeouts. Default is 10
   *   seconds.
   * @param {number} options.sessionExpire[=900] -
   *   Time(seconds) before a connection with a session id expires. Default is
   *   15 minutes.
   * @param {number} options.sessionExpireOnResume[=900] -
   *   The new expiry time(seconds) to set on the connection when a session is
   *   resumed. Default is 15 minutes.
   * @param {number} options.expireInterval[=120] -
   *   Time interval to expire connections.
   * @param {function} options.isValidSession -
   *   A function for checking if a DDP proxy session id is valid. By default
   *   it is a function that always return true.
   */
  configure: function (options) {
    // Don't handle if options is not an object
    if (typeof options !== 'object') return undefined;

    var self = this;
    var defaultOpt = {
      url: Meteor.absoluteUrl(),
      connectionExpire: 900, // 15 minutes
      connectionTimeout: 10,
      sessionExpire: 900, // 15 minutes
      sessionExpireOnResume: 900, // 15 minutes
      expireInterval: 120, // 2 minutes
      ddpConnection: {}, // Options for DDP connection
      // User-defined function to check validity of session ids
      isValidSession: function () { return true; }
    };
    self._config = _.defaults(options, defaultOpt);

    // Restart expire handler with new interval
    self.startExpire();

    return self;
  },

  /**
   * Creates a DDP ProxyConnection to the given server url and login options
   *
   * @param {object} options
   * @param {string} [options.url] - The connection url
   * @param {*} [options.sessionId] -
   *   Session id which is used to resume this connection. A session id must be
   *   a string or a serializable object.
   * @param {object} [options.login] - Login options for Meteor
   * @param {object} [options.login.resume] - Resume token for Meteor login
   * @param {boolean} options.autoPublish[=false] -
   *   If autoPublish is set to true, the connection will subscribe to
   *   autopublished data.
   */
  connect: function (opt) {
    var self = this;
    opt = opt || {};
    opt.url = opt.url || self._config.url; // Use default url if not specified
    if (typeof opt.url !== 'string') throw new Meteor.Error('Invalid url');
    if (typeof opt.login !== 'object') opt.login = {};
    var connection;

    // Attempt to resume session
    if ('sessionId' in opt) {
      if (!self._config.isValidSession(opt.sessionId)) return undefined;
      if (typeof opt.sessionId !== 'string')
        opt.sessionId = EJSON.stringify(opt.sessionId, {canonical: true});

      connection = self._resumeConnection(
        opt.url, opt.sessionId, opt.login.resume
      );
      if (connection) return connection;
    }

    // Create new connection
    return self._addConnection(opt);
  },

  /**
   * Start monitoring connections expiry and remove expired connections.
   */
  startExpire: function () {
    var self = this;
    var interval = self._config.expireInterval * 1000; // Convert to ms
    if (interval < 0 || !isFinite(interval)) interval = 60000;

    self.stopExpire(); // Stop existing interval

    self._expireIntervalId = Meteor.setInterval(function () {
      self._removeExpiredConnections();
    }, interval);

    return self;
  },

  /**
   * Stop monitoring connections expiry
   */
  stopExpire: function () {
    var self = this;
    self._expireIntervalId && Meteor.clearInterval(self._expireIntervalId);
    self._expireIntervalId = null;
    return self;
  },

  /**
   * Close all connections and stop monitoring
   */
  stop: function () {
    var self = this;
    self.stopExpire();
    self._cln.find().forEach(function (doc) {
      var conn = self._connections[doc._id];
      if (conn) conn.close();
    });
  }
});
